File: RenetikObjc/Library/Classes/Core/Categories/UIKit/UIImage+CSExtension.m
//
//  Created by Rene Dohan on 10/14/12.
//

#import "UIImage+CSExtension.h"
#import "CSLang.h"

@implementation UIImage (CSExtension)

- (UIImage *)scaleAndRotateFromCamera :(int)maxResolution {
    CGImageRef imgRef = self.CGImage;
    CGFloat width = CGImageGetWidth(imgRef);
    CGFloat height = CGImageGetHeight(imgRef);
    CGAffineTransform transform;
    CGRect bounds = CGRectMake(0, 0, width, height);
    if (width > maxResolution || height > maxResolution) {
        CGFloat ratio = width / height;
        if (ratio > 1) {
            bounds.size.width = maxResolution;
            bounds.size.height = roundf(bounds.size.width / ratio);
        } else {
            bounds.size.height = maxResolution;
            bounds.size.width = roundf(bounds.size.height * ratio);
        }
    }

    CGFloat scaleRatio = bounds.size.width / width;
    CGSize imageSize = CGSizeMake(CGImageGetWidth(imgRef), CGImageGetHeight(imgRef));
    CGFloat boundHeight;
    UIImageOrientation orient = self.imageOrientation;
    switch (orient) {
        case UIImageOrientationUp : //EXIF = 1
            transform = CGAffineTransformIdentity;
            break;

        case UIImageOrientationUpMirrored : //EXIF = 2
            transform = CGAffineTransformMakeTranslation(imageSize.width, 0.0);
            transform = CGAffineTransformScale(transform, -1.0f, 1.0);
            break;

        case UIImageOrientationDown : //EXIF = 3
            transform = CGAffineTransformMakeTranslation(imageSize.width, imageSize.height);
            transform = CGAffineTransformRotate(transform, M_PI);
            break;

        case UIImageOrientationDownMirrored : //EXIF = 4
            transform = CGAffineTransformMakeTranslation(0.0, imageSize.height);
            transform = CGAffineTransformScale(transform, 1.0, -1.0f);
            break;

        case UIImageOrientationLeftMirrored : //EXIF = 5
            boundHeight = bounds.size.height;
            bounds.size.height = bounds.size.width;
            bounds.size.width = boundHeight;
            transform = CGAffineTransformMakeTranslation(imageSize.height, imageSize.width);
            transform = CGAffineTransformScale(transform, -1.0f, 1.0);
            transform = CGAffineTransformRotate(transform, 3.0 * M_PI / 2.0);
            break;

        case UIImageOrientationLeft : //EXIF = 6
            boundHeight = bounds.size.height;
            bounds.size.height = bounds.size.width;
            bounds.size.width = boundHeight;
            transform = CGAffineTransformMakeTranslation(0.0, imageSize.width);
            transform = CGAffineTransformRotate(transform, 3.0 * M_PI / 2.0);
            break;

        case UIImageOrientationRightMirrored : //EXIF = 7
            boundHeight = bounds.size.height;
            bounds.size.height = bounds.size.width;
            bounds.size.width = boundHeight;
            transform = CGAffineTransformMakeScale(-1.0f, 1.0);
            transform = CGAffineTransformRotate(transform, M_PI / 2.0);
            break;

        case UIImageOrientationRight : //EXIF = 8
            boundHeight = bounds.size.height;
            bounds.size.height = bounds.size.width;
            bounds.size.width = boundHeight;
            transform = CGAffineTransformMakeTranslation(imageSize.height, 0.0);
            transform = CGAffineTransformRotate(transform, M_PI / 2.0);
            break;

        default :
            [NSException raise :NSInternalInconsistencyException format :@"Invalid image orientation"];
    }

    UIGraphicsBeginImageContext(bounds.size);

    CGContextRef context = UIGraphicsGetCurrentContext();

    if (orient == UIImageOrientationRight || orient == UIImageOrientationLeft) {
        CGContextScaleCTM(context, -scaleRatio, scaleRatio);
        CGContextTranslateCTM(context, -height, 0);
    } else {
        CGContextScaleCTM(context, scaleRatio, -scaleRatio);
        CGContextTranslateCTM(context, 0, -height);
    }

    CGContextConcatCTM(context, transform);

    CGContextDrawImage(UIGraphicsGetCurrentContext(), CGRectMake(0, 0, width, height), imgRef);
    UIImage *imageCopy = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();

    return imageCopy;
}

- (UIImage *)scaleToWidth :(float)width {
    float oldWidth = self.size.width;
    float scaleFactor = width / oldWidth;
    float newHeight = self.size.height * scaleFactor;
    float newWidth = oldWidth * scaleFactor;
    return [self getImage :newHeight newWidth :newWidth];
}

- (UIImage *)getImage :(float)newHeight newWidth :(float)newWidth {
    UIGraphicsBeginImageContext(CGSizeMake(newWidth, newHeight));
    [self drawInRect :CGRectMake(0, 0, newWidth, newHeight)];
    UIImage *newImage = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    return newImage;
}

- (UIImage *)scaleToHeight :(float)height {
    float oldHeight = self.size.height;
    float scaleFactor = height / oldHeight;
    float newWidth = self.size.width * scaleFactor;
    float newHeight = oldHeight * scaleFactor;
    return [self getImage :newHeight newWidth :newWidth];
}

+ (UIImage *)imageWithColor :(UIColor *)color {
    CGRect rect = CGRectMake(0.0f, 0.0f, 1.0f, 1.0f);
    UIGraphicsBeginImageContext(rect.size);
    CGContextRef context = UIGraphicsGetCurrentContext();

    CGContextSetFillColorWithColor(context, [color CGColor]);
    CGContextFillRect(context, rect);

    UIImage *image = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    return image;
}

- (UIImage *)inverseColor {
    CIImage *coreImage = [CIImage imageWithCGImage :self.CGImage];
    CIFilter *filter = [CIFilter filterWithName :@"CIColorInvert"];
    [filter setValue :coreImage forKey :kCIInputImageKey];
    CIImage *result = [filter valueForKey :kCIOutputImageKey];
    return [UIImage imageWithCIImage :result];
}

- (UIImage *)template {
    return [self imageWithRenderingMode :UIImageRenderingModeAlwaysTemplate];
}

@end

File: Sources/RenetikCore/extensions/UIKit/UIImage+CSExtension.swift
//
// Created by Rene Dohan on 9/22/19.
//

import UIKit

public extension UIImage {

    convenience init(_ fileName: String) {
        self.init(named: fileName)!
    }

    func imageWith(color: UIColor) -> UIImage {
        let rect = CGRect(x: 0.0, y: 0.0, width: 1.0, height: 1.0)
        UIGraphicsBeginImageContext(rect.size)
        let context = UIGraphicsGetCurrentContext()!
        context.setFillColor(color.cgColor)
        context.fill(rect)
        let image = UIGraphicsGetImageFromCurrentImageContext()!
        UIGraphicsEndImageContext()
        return image
    }

    func inverseColor() -> UIImage {
        let coreImage = CIImage(cgImage: cgImage!)
        let filter = CIFilter(name: "CIColorInvert")!
        filter.setValue(coreImage, forKey: kCIInputImageKey)
        let result = filter.value(forKey: kCIOutputImageKey) as! CIImage
        return UIImage(ciImage: result)
    }

    var template: UIImage {
        withRenderingMode(.alwaysTemplate)
    }

    func scale(width: CGFloat) -> UIImage {
        let oldWidth = size.width
        let scaleFactor = width / oldWidth
        return scale(width: oldWidth * scaleFactor, height: size.height * scaleFactor)
    }

    func scale(height: CGFloat) -> UIImage {
        let oldHeight = size.height
        let scaleFactor = height / oldHeight
        return scale(width: size.width * scaleFactor, height: oldHeight * scaleFactor)
    }

    func scale(width newWidth: CGFloat, height newHeight: CGFloat) -> UIImage {
        UIGraphicsBeginImageContext(CGSize(width: newWidth, height: newHeight))
        draw(in: CGRect(width: newWidth, height: newHeight))
        let newImage = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()
        return newImage!
    }

    func scaleAndRotateFromCamera(_ maxResolution: CGFloat) -> UIImage {
        let imgRef = cgImage!
        let width = CGFloat(imgRef.width)
        let height = CGFloat(imgRef.height)
        var bounds = CGRect(x: 0, y: 0, width: width, height: height)
        if width > maxResolution || height > maxResolution {
            let ratio = width / height
            if ratio > 1 {
                bounds.size.width = maxResolution
                bounds.size.height = CGFloat(roundf(Float(bounds.size.width / ratio)))
            } else {
                bounds.size.height = maxResolution
                bounds.size.width = CGFloat(roundf(Float(bounds.size.height * ratio)))
            }
        }
        let scaleRatio = bounds.size.width / width
//        let imageSize = CGSize(width: width, height: height)
        var transform: CGAffineTransform? = nil

        switch imageOrientation {
        case .up /*EXIF = 1 */:
            transform = CGAffineTransform.identity
        case .upMirrored /*EXIF = 2 */:
            transform = CGAffineTransform(translationX: width, y: 0.0).scaledBy(x: -1.0, y: 1.0)
        case .down /*EXIF = 3 */:
            transform = CGAffineTransform(translationX: width, y: height).rotated(by: .pi)
        case .downMirrored /*EXIF = 4 */:
            transform = CGAffineTransform(translationX: 0.0, y: height).scaledBy(x: 1.0, y: -1.0)
        case .leftMirrored:
            bounds.size.height = bounds.size.width
            bounds.size.width = bounds.size.height
            transform = CGAffineTransform(translationX: height, y: width)
                    .scaledBy(x: -1.0, y: 1.0).rotated(by: 3.0 * .pi / 2.0)
        case .left:
            bounds.size.height = bounds.size.width
            bounds.size.width = bounds.size.height
            transform = CGAffineTransform(translationX: 0.0, y: width).rotated(by: 3.0 * .pi / 2.0)
        case .rightMirrored:
            bounds.size.height = bounds.size.width
            bounds.size.width = bounds.size.height
            transform = CGAffineTransform(scaleX: -1.0, y: 1.0).rotated(by: .pi / 2.0)
        case .right:
            bounds.size.height = bounds.size.width
            bounds.size.width = bounds.size.height
            transform = CGAffineTransform(translationX: height, y: 0.0).rotated(by: .pi / 2.0)
        @unknown default:
            NSException(reason: "Invalid image orientation").raise()
        }

        UIGraphicsBeginImageContext(bounds.size)
        let context = UIGraphicsGetCurrentContext()!
        if imageOrientation == .right || imageOrientation == .left {
            context.scaleBy(x: -scaleRatio, y: scaleRatio)
            context.translateBy(x: -height, y: 0)
        } else {
            context.scaleBy(x: scaleRatio, y: -scaleRatio)
            context.translateBy(x: 0, y: -height)
        }
        context.concatenate(transform!)
        context.draw(imgRef, in: CGRect(x: 0, y: 0, width: width, height: height))
        let imageCopy = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()
        return imageCopy!
    }
}